metrics-threadpool/src/main/java/com/alibaba/metrics/threadpool/ThreadPoolMetricsGaugeSet.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2017 Alibaba.com All right reserved. This software is the
+ * confidential and proprietary information of Alibaba.com ("Confidential
+ * Information"). You shall not disclose such Confidential Information and shall
+ * use it only in accordance with the terms of the license agreement you entered
+ * into with Alibaba.com.
+ */
+package com.alibaba.metrics.threadpool;
+
+import com.alibaba.metrics.*;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This gauge set is used to collect @see ThreadPoolExecutor metrics
+ **/
+public class ThreadPoolMetricsGaugeSet extends CachedMetricSet {
+
+
+    private static final String[] THREAD_POOL_METRICS_GUAGES = new String[]{"active", "queued", "completed", "pool"};
+
+
+    private long[] threadPoolMetrics;
+    private final ThreadPoolExecutor threadPoolExecutor;
+    // use short ttl
+    private static long DEFAULT_DATA_TTL = 10;
+
+    public ThreadPoolMetricsGaugeSet(ThreadPoolExecutor threadPoolExecutor) {
+        this(DEFAULT_DATA_TTL, TimeUnit.MILLISECONDS, Clock.defaultClock(), threadPoolExecutor);
+    }
+
+    public ThreadPoolMetricsGaugeSet(long dataTTL, TimeUnit unit, ThreadPoolExecutor threadPoolExecutor) {
+        this(dataTTL, unit, Clock.defaultClock(), threadPoolExecutor);
+    }
+
+    public ThreadPoolMetricsGaugeSet(long dataTTL, TimeUnit unit, Clock clock, ThreadPoolExecutor threadPoolExecutor) {
+        super(dataTTL, unit, clock);
+        if (threadPoolExecutor == null) {
+            throw new RuntimeException("illegal thread pool executor,must not be null");
+        }
+        threadPoolMetrics = new long[THREAD_POOL_METRICS_GUAGES.length];
+        this.threadPoolExecutor = threadPoolExecutor;
+    }
+
+    @Override
+    protected void getValueInternal() {
+
+        if (threadPoolExecutor == null) {
+            return;
+        }
+        // active count
+        threadPoolMetrics[0] = threadPoolExecutor.getActiveCount();
+        // queued size
+        threadPoolMetrics[1] = threadPoolExecutor.getQueue().size();
+        // completed count
+        threadPoolMetrics[2] = threadPoolExecutor.getCompletedTaskCount();
+        // pool size
+        threadPoolMetrics[3] = threadPoolExecutor.getPoolSize();
+    }
+
+    @Override
+    public Map<MetricName, Metric> getMetrics() {
+        Map<MetricName, Metric> metrics = new HashMap<MetricName, Metric>();
```

Comment:
I still think this can be improved. When you call getMetrics() every time, a new HashMap can a few ThreadPoolGauge will be created, which is unnecessary. I think they can be created in the constructor of `ThreadPoolMetricsGaugeSet`.  How do you think?

Reply:
It's reasonable! Already change it !